### ISSUE TITLE
Option to delete and/or edit own logs

### ISSUE THREAD

**Issue by jeancaffou** [cgeo/cgeo#2012]
Sometimes when typing, the sneaky keyboard hides and you press "send log" accidentially.

**Comment by Lineflyer** [comment]
This is the same as the ability to edit logs in general which we declined for now. Refer to mentioned issues.

**Comment by Lomanic** [comment]
I couldn't find anywhere in the mentioned issues why this was declined. This is elementary and lacking since forever, we shouldn't have to open a browser, go to the cache listing, find our log, click the "Upload image/edit log" link, modify the text and send the form (and then, reloading the cache data locally as described in https://github.com/cgeo/cgeo/issues/8888#issuecomment-681129322). Mistakes happen in logs in cgeo and it's not unusual to want to fix them.

**Comment by fm-sys** [comment]
I agree, this should be reopened. There is no real reason, why this shouldn't be integrated in c:geo if someone would be interested in implementing this sometime...

**Comment by eddiemuc** [comment]
I would tackle this issue, but only after #2553 / #8822 is finalized, if that's all right

**Comment by eddiemuc** [comment]
#8822 is merged (and first bugs already coming in...) so time to think about this one.

I suggest to approach this as follows and would be eager to have some feedback:

**General idea**:
- place "edit"/"delete" icon buttons near those logbook entries which user is allowed to edit
- for edit: Reuse (offline) log edit dialog to "load in" existing logs, then edit and resend them

**MVP**:
- start with geocaching.com
- keep "one-offline-log-per-cache" policy for now: if log should be edited while for same cache an offline log already exists, existing offline log is deleted (after questioning user of course)
- add some sort of hint/label to offline log dialog when editing existing entry. Allow conversion from "editing existing log entry" to new log entry

**Extensions/final solution**:
- add feature also for opencaching
- extend c:geo GUI such that multiple offline log entries can be stored per cache ("edit" as well as "new")
- allow refresh of currently edited log entry

That would be the rough idea. What do you think?

**Comment by Lomanic** [comment]
> place "edit"/"delete" icon buttons near those logbook entries which user is allowed to edit

I think that would clutter the UI if it's always displayed. I would have seen this implemented using the menu displayed on long press on a log entry (where one can copy/share or translate log's text currently).

**Comment by Lineflyer** [comment]
We could even discuss to swap current long and short click.

Currently short click performs ROT13. Its IMHO rarely useful and could either go into long press or into a common menu behind short click together with edit/delete.

**Comment by eddiemuc** [comment]
Hi @Lomanic and @Lineflyer I agree to both of your comments.

However, I want some sort of small visible hint visible in the log list whether a log entry is editable to the user or not. What about simply adding the word "editable" (translated to selected language of course) in the left column of a log entry (between log type and the number of founds of the user wroting the entry).

E.g. a bit like this:
![image](https://user-images.githubusercontent.com/6909759/92203623-e184f380-ee81-11ea-92ad-df0728f6b78f.png)

**Comment by moving-bits** [comment]
@eddiemuc 
Good idea, but I would like to suggest a slightly different placement: Either in the same row as proposed, but as last entry (to distinguish it from the already existing information), or in the same line as the geocacher's name, but right-aligned (and with the same small font as proposed).

**Comment by eddiemuc** [comment]
Ok

**Comment by Lomanic** [comment]
I find that this is cluttering the UI, why is it needed to display if a log is editable? Only the player's own logs are editable (on GC at least), so this is redundant with the player's name. For cache owners, they can delete (not edit) any log on their cache, and I'm quite sure they are aware of which caches they own. 

Compared to Octodroid (again I know, but the app is not 10 years old and has honestly better UX, while doing quite a few things), it is not displayed that you can edit/delete your own comments (or comments on your projects), but there is a three dot menu on each comment and its Edit and Delete entries are only added on comments you can edit/delete. That could be the same thing on cgeo (that's how I was seeing this at least), on long press on a log or on tap as was suggested earlier. 

If you really want to give a visual clue that a log is editable, I would like to suggest something more subtle, like as small pencil (though it could be confused as hinting that the log has been modified, this is what is done in Octodroid).

**Comment by eddiemuc** [comment]
Hi @Lomanic , thank you for your comment. I took away two things about it:
* I did not plan until now to offer deletion of logs by owners in c:geo. The reason is that c:geo does not seem to focus on tools for owners up to now, and this would kind of break with that convention. Do you think that should be offered?
* My wish fora visual hint originated from the geocaching.com website. Here in the log list you see a "View / Edit Log / Images   Upload Image" on your own logs and a "View Log" on other people's logs. Strangely enough, when you view a cache that you own, you still just see "View Log" links on other people's logs (while you can delete them after clicking on that link). But the geocaching.com website might not be the best place to look for a good UI anyway. So I checked both the geocaching app and Geooh Go (the only other apps I have installed). And you are right, there is no special visual clue whether a log can be edited or not. -> So I would move away from my original opinion and NOT include a visual clue

**Comment by Lomanic** [comment]
The ability to delete logs on owned caches can be postponed or not implemented even, this use case is quite rare compared to fixing a typo in one's own logs.

But I don't think adding this feature will complicate the code much: it's the same logic (HTTP calls) for any log you can delete (own log or not) and the menu entry still has to be populated under some conditions (testing if selected log's author is the player ; cgeo already has to have code for this to be able to post owner-specific types of logs on owned caches). In pseudocode it only modifies a few lines:

```
// edit/delete only own logs
if (log.author == player) {
    menu.add("Edit")
    menu.add("Delete")
}

// edit/delete own logs and delete any log on owned cache
if (log.author == player) {
    menu.add("Edit")
}
if (log.author == player || cache.owner == player) {
    menu.add("Delete")
}
```

The rest of the implementation wouldn't change.

**Comment by eddiemuc** [comment]
Note to myself with regards to how to parse/set values for geocaching.com: valuable input in this Pull Request #6686 https://github.com/cgeo/cgeo/pull/6686

**Comment by eddiemuc** [comment]
It seems to me I have to restore some of the code removed by @pstorch in 2017 because the GC Logging pages for edit/remove are still the same as before the big logging change. Also I wonder how @pstorch was able to refactor some of the API details from the web page

@pstorch I could really use some talking time with you:
* Is my guess correct that editing logs might only be possible by using the "old" `ctl00$ContentBody$LogBookPanel1`-type-parameters?
* How did you find out the correct sequence to upload images using the new logging pages? (I need to find out how to replace logging images in existing logs, this is why I am asking)

**Comment by pstorch** [comment]
If they still use the old page for editing it would be a maintenance disaster for c:geo to implement both ways. I was glad to be able to remove old code from c:geo back then. But it's up to you to do it.

Regarding how I found out: I used the web developer tools in the browser and recorded the network traffic while logging. Then I could see what the GC webpage was sending and receiving.

**Comment by eddiemuc** [comment]
@pstorch thanks for the insights! And you are right with regards to the "maintenance disaster". I don't understand myself why GC decided not to touch edit/delete/view-log pages but only create-page. I have to think about that...

**Comment by fm-sys** [comment]
> it would be a maintenance disaster for c:geo to implement both ways

are there still any UI changes related to the old logging-page? I would guess that GC will not develop it further more. So is there really much maintenance work to be expected?

**Comment by geocachermgo** [comment]
Maybe the inconsistency between menu point "German Leeren" (english: empty?) and can also be solved. If I remove all the text I still have an (empty) log. When I found use the menu ("Leeren") the log is deleted.

**Comment by eddiemuc** [comment]
In case you wonder why there is no progress: I will continue in this one once #8897 is finished (because I want to build up on the work done there)

**Comment by eddiemuc** [comment]
Since SAF is currently on hold, this would be the next issue on my list.

My idea would be to implement this in a first step as follows:
* For gc.com only (but preparing the connector structures such that other connectors can be implemented later)
* Add a new entry in the log entry context menu named "Edit Log Entry" (only for own entries and if necessary information is available e.g. the connector-specific log id)
* This will open the "Offline log edit" activity with the locally saved log state prefilled, 
* For now I would restrict possible edits to date, log type and description. Esp in the first step I would not implement image edit (the rest is already complicated enough and imho already brings user benefit)
* If user edits but discards entries I would also discard the edited changes (I would NOT store this as a new "offline log"; this would open up a whole new area of complexity)

Any opinions about that?

**Comment by fm-sys** [comment]
Sounds good! Only image edit (mainly add more pictures but maybe also delete others) would be quite nice to have. Would it be really that difficult to implement? 

And you are right, edit is mostly used for some small quick fixes and therefore no need for storing them offline...

**Comment by moving-bits** [comment]
> Any opinions about that?

LGTM, thanks!

Regarding editing log images: Would be nice to have, but ok to implement in step 2, IMHO.

**Comment by eddiemuc** [comment]
Regarding image edit: I agree that this would be nice to have. If during coding it turns out that this is easy to include I will do it. But looking at how this is done on the GC website I assume it is much more complex than the rest. Also, I observed that even geooh GO does not allow this.

**Comment by eddiemuc** [comment]
I thought about this topic over night.

Maybe the deep Integration of this into c:geo is not worth the effort. Maybe it is enough to offer a link to the "edit log" page in the browser (which should be rather easy since we already have a link to open the log in browser in read-only-mode). And maybe a second link to delete a log entry.

Why am I thinking this?
* The deep Integration into c:geo will be very complicated, esp when also including images. Due to GC.com using different pages and methods on their website to create and edit logs, this means re-importing back and maintaining a whole lot of code which is just used for this function (see comments of @pstorch on this)
* On the other hand this function will not be used very often, and it can only be used while online anyway. Please be honest with yourself: when was the last time you needed to edit log images for a log entry while not sitting before a browser (e.g. having to do this via c:geo)

Given the above, my feeling is that we have more important stuff on our plate than a deep Integration. What do you think?

On other notes:
* Unfortunately it turned out that I didn't find the time yesterday to work on this
* I also remembered that I started working on a migration to mapsforge 0.14 for whereyougo which I would rather continue before starting the log edit thingy

**Comment by fm-sys** [comment]
Probably you are right. For the GC message center, we have also decided against a deep integration and are only offering a link to the browser page... The complicated way is not automatically the better solution!

**Comment by moving-bits** [comment]
> Maybe the deep Integration of this into c:geo is not worth the effort. Maybe it is enough to offer a link to the "edit log" page in the browser.

I'm not gonna stop you either way, but IMHO you're right,

> I also remembered that I started working on a migration to mapsforge 0.14 for whereyougo which I would rather continue before starting the log edit thingy

especially given this alternative. Given the amount of work needed for implementing an "edit log" workflow, this could rather be directed for a more widely-used activity, and bringing the WIG mapsforge support to "state-of-the-art" seems like a good choice for this.

Personally I use logging from c:geo only rareley - actually only in the rare case, that I'm doing some geocaching, but don't have my PC/laptop available for at least a week or so.

**Comment by geocachermgo** [comment]
Since the link to a single log is already available the advantage to save one click to access the edit page is only small.
On the contrary to moving-bit I do increasing number of logs at the phone and I had several conversations with newbies that only use an app and did'nt knew that it is possible to edit logs.
But I agree that it's a lower ranking feature.

**Comment by alan666notb** [comment]
An option I would like is simply to add an image to an existing log I just submitted (Cos I forgot).

Doing it via 'open in browser' while out out in the field is surprisingly difficult - the file selector in Chrome always seems to make it difficult for me, and the image doesn't get re-sized and Exif-stripped before uploading.

**Comment by fm-sys** [comment]
> Maybe the deep Integration of this into c:geo is not worth the effort. Maybe it is enough to offer a link to the "edit log" page in the browser (which should be rather easy since we already have a link to open the log in browser in read-only-mode).

There was a similar opinion in #5170:
> I've been asked multiple times of log editing is possible. While I don't want to implement that directly in the app, we could still link to the browser page for editing.

if someone wants to implement it anyway, it should be done for OC as well, see #7241

**Comment by eddiemuc** [comment]
Option to open log in browser is already implemented for both oc and GC ("open in browser" in context menu). It is one more click for the user there to edit it (if it is his own log of course). Is it really worth the effort to further dig into this?

**Comment by ztNFny** [comment]
> Is it really worth the effort to further dig into this?

Imho it is - for the same reason that c:geo offers a logging UI in the first place instead of just linking to the website:
- The website is not mobile-optimized at all, making it quite cumbersome to use in a mobile browser. 
- The editor text size on the website will either be tiny or you have to horizontally-scroll all the time.
- and of course it's loading a lot more resources than required (in terms of network traffic)

For me the main feature about "edit log" is editing the log text (as opposed to images), e.g. when I made a typo that I noticed too late.

**Comment by anondev-sudo** [comment]
2012 and it still wasn't added such important feature

**Comment by ztNFny** [comment]
> 2012 and it still wasn't added such important feature

You're welcome to contribute!

**Comment by eddiemuc** [comment]
> 2012 and it still wasn't added such important feature

This is a very complex feature to add. Up until recently it was not really reachable. Recent changes on gc.com side would make it at least possible now. But still this is a major piece of work.

**Comment by eddiemuc** [comment]
With PR #15052 merged, an experimental and not-yet-finished version of log edit is now available in the nightlies. To test it you need to add a boolean Setting `feature_log_edit` and set it to `true` (use `Settings -> System -> View Settings -> Edit -> Add`)

This will add two new context menu entries "edit log" and "delete log") when tapping on a log entry (only if the log entry supports edit/delete log):

![image](https://github.com/cgeo/cgeo/assets/6909759/ca34a1a7-d2fa-4961-b2c7-a5655b4c1770)


If tapping on "Edit log" the following dialog will appear (as of now with "warning" label on top):

![image](https://github.com/cgeo/cgeo/assets/6909759/5b8832fc-ae93-4dcf-9652-38a65b65b854)

Editing the entries and tapping "send" will send the modified log to server.

The code is still very experimental and not well tested. The following restricitons are already known:
* Feature is not end-user-styled and not internationalized. Popping-up dialog texts are not translated and not yet user-friendly
* Feature is only implemented for gc.com yet. gc.com allows edit/delete for own logs and delete for logs of own caches 
* Edit works for text, log type and log date
* Image edit is still restricted: Log images can be added and deleted. However existing images are not shown yet in the image list. Image property edit (title, description) is not tested yet. Editing the image itself won't be supported by the API and must still be forbidden in the GUI
* Not tested whether log edit/delete works for caches not locally stored in DB
* ...probably more problems will come up

Regardless of the many restrictions I share this feature with already you because I think even in this state it might help out in the field. Merry christmas.

**Comment by geocachermgo** [comment]
I edit often so I like this feature. Just tried to edit an existing find log. There was a new "Bemerkung/Note" Log created instead of an edit to the old one.

**Comment by eddiemuc** [comment]
> I edit often so I like this feature. Just tried to edit an existing find log. There was a new "Bemerkung/Note" Log created instead of an edit to the old one.

Hi @geocachermgo thanks for the fast feedback :-) I just tried to edit a "Found" log of myself for the first time. I noted indeed that c:geo doesn't offer the "Found" log type in this case (-> needs to be changed, I believe c:geo currently doesn't offer "Found" log type on an already found cache to avoid "Found"-log-duplicates). But it allowed me to change  my "found" log to a "write note", it didn't create a new log entry. Did you really experience a new log being created in addition to the existing one?

**Comment by geocachermgo** [comment]
> Did you really experience a new log being created in addition to the existing one?
Yes (with yesterdays nightly), the edit button on a found log created the same behavior in c:geo and at the geocaching site as the "Log visit" from the menu:
*New entry with type remark that can only be changed to (DNF, Note, NM, NA) 
*Automatic text "Danke für den Cache, mgo" filled in
*Two entries in c:geo Loglist
*With "Log leeren/empty logs" it dissappears
*If created again and sent to the website there is a new remark entry.
If I  use edit for the new remark and choose delete, that log is deleted on the website, but still there on c:geo until reload.

**Comment by geocachermgo** [comment]
With todays nightly it worked:
* Old logtype (found) date kept
Changed on the website
Strange success report:
![Screenshot_2023-12-22-07-49-56-813_cgeo geocaching](https://github.com/cgeo/cgeo/assets/43813636/e099bc3a-bb89-4bec-bd69-315d4871212a)
* Bug: Two entries in cgeo loglist, new (marked orange as not sent) is not deleted by reload. On the website there is only one log.
![Screenshot_2023-12-22-07-50-09-426_cgeo geocaching](https://github.com/cgeo/cgeo/assets/43813636/7638c2a4-153a-4b95-a942-9bcdf43c29ee)
* If I sent it, the type is changed to remark an there are two logs that look the same exept first is "found", second is remark

**Comment by eddiemuc** [comment]
> With todays nightly it worked:
> 
> * Old logtype (found) and date kept
>   Changed on the website
> * Strange success report (ok, since new feature)

Yeah,  GUI is not in a user-friendly state yet :-(

> * Bug: Two entries in cgeo loglist, new (marked orange as not sent) is not deleted by reload. On the website there is only one log.

That one is strange. Probably interaction when an offline log is stored and an existing log is edited doesn't work correctly. I have to look into this. All experimental for now...

> * If I sent the top one, the type is changed to remark and there are two logs that look the same exept second is "found", first (newer) is "remark" both in c:geo and on the website.

Given the bug above, this is a "follow-up" problem. I believe c:geo does not allow to send a second "Found" log when a "Found" log already exists, I assume this is why on sending the "top" one it is converted to a "write note" under the hood. The whole log logic of c:geo currently is meant for new logs only, and this entangles in strange ways with editing of logs, it will take a while to untabgle all of this...

**Comment by eddiemuc** [comment]
@geocachermgo your last reported problems should be fixed with PR #15078. Thanks btw for your testing of this feature in this very early phase, the feedback is really helpful!

**Comment by geocachermgo** [comment]
Always happy to help. Do you see any possible connection with #15068 ?

**Comment by eddiemuc** [comment]
> Always happy to help. Do you see any possible connection with #15068 ?

I don't believe this is connected to the log edit/delete feature. But there is a lot going on with the new unified map in parallel. I assume it is related to that

**Comment by geocachermgo** [comment]
> > Always happy to help. Do you see any possible connection with #15068 ?
> 
> I don't believe this is connected to the log edit/delete feature. But there is a lot going on with the new unified map in parallel. I assume it is related to that

Yes, the #15034 was created earlier than the log edit option.

**Comment by eddiemuc** [comment]
> This will add two new context menu entries "edit log" and "delete log") when tapping on a log entry (only if the log entry supports edit/delete log):
> 
> ![image](https://private-user-images.githubusercontent.com/6909759/292160868-ca34a1a7-d2fa-4961-b2c7-a5655b4c1770.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDQzOTY0MTQsIm5iZiI6MTcwNDM5NjExNCwicGF0aCI6Ii82OTA5NzU5LzI5MjE2MDg2OC1jYTM0YTFhNy1kMmZhLTQ5NjEtYjJjNy1hNTY1NWI0YzE3NzAucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDEwNCUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDAxMDRUMTkyMTU0WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9ZTcxYTNiOGFkZjc3MGVhOWEyYWYzMjQxNDllNDAyYmY1ZTZlM2JhYmVmN2ZhN2EwYTE5NjIzNWZkMmJkM2RkNCZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.qKaaaSw7I96X8AQ0UAmWxSUdbZcGlvHdA7kibNiGnoE)
> 
> If tapping on "Edit log" the following dialog will appear (as of now with "warning" label on top):
> 
> ![image](https://private-user-images.githubusercontent.com/6909759/292160991-5b8832fc-ae93-4dcf-9652-38a65b65b854.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDQzOTY0MTQsIm5iZiI6MTcwNDM5NjExNCwicGF0aCI6Ii82OTA5NzU5LzI5MjE2MDk5MS01Yjg4MzJmYy1hZTkzLTRkY2YtOTY1Mi0zOGE2NWI2NWI4NTQucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDEwNCUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDAxMDRUMTkyMTU0WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9MjcyMGQ0ZmFhZjIyYjliODA2MWYwOGMxM2Y2NjZhMWEwNDFlNWIyODY5YTZhYWUwZDYyN2Y0OGY3NWE1YmMwZSZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.SDFebux5nPgRRq8A_Y5cKf6515uQ2k4H4m7ASt_RmC8)
> 
> Editing the entries and tapping "send" will send the modified log to server.
 
PR #15163, if merged, concludes the implementation of this feature. Since this issue is already far too long, I would ask for additional requests and bug reports to be filed as new issues.

The feature is for now limited to gc.com. However, the connector mechanics are in place to enable other platforms as well. Please file separate issues if this is wanted.

**Comment by anondev-sudo** [comment]
@eddiemuc 
Btw there needs to be added length verification to delete reason field, because when you write longer text, you will get error which doesn't say why the log deletion was not successful. Display error if the length of the reason exceeds 125 characters. I was confused why I couldn't delete log but after using the web version I found out that there is 125 char limit set to reason field

![image](https://github.com/user-attachments/assets/167cacfb-ee4e-4f60-9c25-85d91b80886d)
